Eeprom data Structure

saching13 opened the pull request:
This is just the basic idea I had for scenarios when there can be many cameras. in that situation. users have the flexibility to load and associate this calibration to any cameras(needs some changes). 
Things I am not sure about is whether we can use pointers like that?
should I replace it with camid / something else ?  

In gen1 we had `left_to_rgb distance` and `left_to_right distance`. if there are multiple cameras, or in the custom board, it doesn't make sense to use left or right since they can use any two cameras for stereo or any other tasks. So, I was thinking we should just mention it in extrinsic adding one more parameter as `measured_translation` or something like that. 

What do you guys think ? and what would be better to make it flexible like gen2.

themarpe commented on include/depthai-shared/common/EepromData.hpp:
```diff
@@ -0,0 +1,85 @@
+struct CameraExtrinsics {
+    std::vector<std::vector<double>> rotationMatrix;
+    std::vector<double> translation;
+    std::shared_ptr<CameraInfo> destCamera; // Does using pointer makes sense while writing to eeprom or should we just use an number index for cameras and link using that ?
```
I'd suggest using CameraBoardSocket as "id"

themarpe commented on include/depthai-shared/common/EepromData.hpp:
```diff
@@ -0,0 +1,85 @@
+
+struct StereoRectification {
+    std::vector<std::vector<double>> rectifiedRotationLeft, rectifiedRotationRight;
+    std::shared_ptr<CameraInfo> *leftCamera, *rightCamera;
```
Here as well

themarpe reviewed (changes requested):
Above remarks as well as there seems to be a merge issue.

saching13 commented:
@themarpe changed it to use CameraBoardSocket and added an extra var to `CameraBoardSocket` enum. and used it as the length of the camera rig. And also added a measured_translation variable. 

Another thing I wanted to discuss was fov variables. 
One thought I had was moving them inside the cameraInfo incase the custom boards might have different camera with different fov. 
Second was do we need to save it. it can be calculated using other intrinsics matrix right ? Does this apply to alll cameras for ex: wide/fisheye ? cc: @alex-luxonis

themarpe commented on include/depthai-shared/common/CameraBoardSocket.hpp:
```diff
@@ -8,6 +8,6 @@ namespace dai {
  * AUTO denotes that the decision will be made by device
  */
-enum class CameraBoardSocket : int32_t { AUTO = -1, RGB, LEFT, RIGHT };
+enum class CameraBoardSocket : int32_t { AUTO = -1, RGB, LEFT, RIGHT, END };
```
I think we can go without `END` here and change the array to unordered map, as this is only data presentation.

themarpe replied:
Or if really needed, I'd rather see `COUNT` or something a bit more out of context

themarpe commented on include/depthai-shared/common/EepromData.hpp:
```diff
@@ -0,0 +1,89 @@
+    bool swapLeftRightCam;
+    double monoFovRad, rgbFovRad;
+    std::string name, revision;
+    std::array<CameraInfo, CameraBoardSocket::END> cameraData;
```
`std::unordered_map<CameraBoardSocket, CameraInfo> cameraData`

themarpe commented on include/depthai-shared/common/EepromData.hpp:
```diff
@@ -0,0 +1,89 @@
+
+
+struct EepromData {
+    int16_t camera_rig_size = 3; // or we can use this for size of the array in cameraData 
```
Is this suppose to present connected cameras? If so then we cannot be sure that they'll follow order.
Lets just deduct this size from `cameraData` map

themarpe commented on include/depthai-shared/common/EepromData.hpp:
```diff
@@ -0,0 +1,93 @@
+
+namespace dai {
+
+struct point {
```
Create a `common/Point3f.hpp`, with `struct Point3f { float x,y,z; };`

themarpe commented on include/depthai-shared/common/EepromData.hpp:
```diff
@@ -0,0 +1,93 @@
+    // (x, y, z) pose of destCameraSocket w.r.t currentCameraSocket obtained through calibration
+    point translation;
+    // (x, y, z) pose of destCameraSocket w.r.t currentCameraSocket measured through CAD design
+    point measured_translation;
```
Use `camelCase`

themarpe commented on include/depthai-shared/common/EepromData.hpp:
```diff
@@ -0,0 +1,93 @@
+struct StereoRectification {
+    std::vector<std::vector<double>> rectifiedRotationLeft, rectifiedRotationRight;
+    CameraBoardSocket leftCameraSocket, rightCameraSocket;
+    int baseline;
```
Should `baseline` rather be float?

themarpe commented on include/depthai-shared/common/EepromData.hpp:
```diff
@@ -0,0 +1,93 @@
+struct EepromData {
+    uint32_t version;
+    bool swapLeftRightCam;
+    std::string board_name, board_rev;
```
Use `camelCase`

alex-luxonis commented on include/depthai-shared/common/EepromData.hpp:
```diff
@@ -0,0 +1,95 @@
+};
+
+struct StereoRectification {
+    std::vector<std::vector<float>> rectifiedRotationLeft, rectifiedRotationRight;
```
Thinking about being able to do StereoDepth between any cameras (even RGB + Mono), should we move the `rectifiedRotation` matrix to `CameraInfo`? Currently I also don't see the RGB rotation matrix specified.

saching13 replied:
RGB doesn't have rectification rotation. 
And rectification is relative to the other camera so it would require us to have rectification between left/right, right/rgb, and left/right making up to 6 matrixes on OBC case and more if we have 4 or more cameras. 
So Stereo Rectification will have for the defaults that we computed. 
And rectification for other can be computed by using [this](https://docs.opencv.org/3.4/d9/d0c/group__calib3d.html#ga617b1685d4059c6040827800e72ad2b6) algorithm. we can recreate this or use it on the host side to pass the obtained rectification, cameras using overload functions like
overleadLeftRectification/ overloadRIghtRectification. 

And as of `baseline`, it seems redundant. will remove it.

saching13 commented on include/depthai-shared/common/EepromData.hpp:
```diff
@@ -0,0 +1,95 @@
+};
+
+struct CameraInfo {
+    std::vector<std::vector<float>> intrinsicMatrix;
```
Yes. I think it would be good to add width and height which will act as a reference for at what res calibration was added. Will add that. Thanks for pointing it out.  
And as of FOV from specs I have already added that for the same reason.

themarpe reviewed (changes requested):
`matrixOps.hpp` shouldn't reside in shared.
Use appropriate library on both FW side and in `core`
Maybe Eigen might be a good one to checkout?: https://hunter.readthedocs.io/en/latest/packages/pkg/Eigen.html

saching13 commented:
MatrixOps contains only matrix multiplication and inverse and these two are used on both firmware and on device side. 
I thought of using Eigen, but it didn't make sense to add the lib just for multiplication and inverse. 
So may be if we start using more functionalities then we can move to Eigen. 

Thoughts ?

themarpe commented:
> MatrixOps contains only matrix multiplication and inverse and these two are used on both firmware and on device side.
> I thought of using Eigen, but it didn't make sense to add the lib just for multiplication and inverse.
> So may be if we start using more functionalities then we can move to Eigen.
> 
> Thoughts ?

You can add it to core and FW separately for now if needed before Eigen or some other library.

The rational is that `depthai-shared` houses only needed data structures (with minimal supporting code) to facilitate communication between FW and host. 
Borderline case is `nanorpc` which would probably land as a separate library in both FW & host if not for many customizations needed to make it work on both ends.

saching13 commented:
Ah okay. My bad. I thought any helper codes that are shared between core and HW can be added here.

alex-luxonis commented on include/depthai-shared/common/EepromData.hpp:
```diff
@@ -0,0 +1,32 @@
+ */
+struct EepromData {
+    uint32_t version = 6;
+    bool swapLeftRightCam = false;
```
About this parameter, currently it is unused. Should we remove it, or if implemented, where should it take effect:
- directly in swapping the LEFT and RIGHT camera board sockets
- inside the StereoDepth node, and then should it be moved to StereoRectification

Also it's a bit confusing as in the existing json configs (e.g [here for OAK-D](https://github.com/luxonis/depthai/blob/eaa15f1/resources/boards/BW1098OBC.json#L6)) is was inverted, due to some historical reasons. 

Thoughts @saching13 @szabi-luxonis @themarpe ?

saching13 replied:
> directly in swapping the LEFT and RIGHT camera board sockets

I like this idea than specifying but it would make it confusing since LEFT and RIGHT is used for naming board sockets. 
It would be good to add alternative naming for them based on which port is connected. For ex: something like CM1, CM2, CM3 etc. 
This would help in better readability of code too when users build custom boards on top of ours or use different Camera Socket on FCC module. 

Thoughts ?

themarpe replied:
I'm up for moving this into a more appropriate place. It also doesn't have to be explicitly mentioned as upgrading old calibrations can just set the calibration data correctly (or swapping them when setting) and this information will be "upgraded".

Maybe an issue is if we don't want to reflash but still move calibration data around? But probably won't happen much to be an issue to just reflash the eeprom. And with current API that should be possible to swap calib data between two sockets right?

saching13 replied:
For V5 calibration already stored I don't think we need to reflash. I can read it left and right and load it appropriately based on swap mode for that instance.

saching13 replied:
and yeah looks like we can remove it from V6

saching13 replied:
cc: @alex-luxonis  @themarpe

alex-luxonis replied:
I'm also inclined towards removing this field. During calibration we can detect if the cameras are swapped and warn the user, and can have an option to calibrate with the cameras swapped. And the swapping can be deduced later in FW from the extrinsics translation vector orientation. I'm just not sure if we should have the user swap the `left` and `right` inputs linked to StereoDepth, or we should do this internally inside the node.

saching13 replied:
Hhmm. 
I am also inclined to remove it. Since I prefer if `CameraSocketBoard` represents a specific port instead of which camera in a stereo setup to avoid more confusion in the future.

if it's okay with everyone I will go ahead and remove it and swap left and right for v5 internally. And during calibration. 
And stereo rectification of calibration part will contain which socket is left and which socket is right which we can validate and throw a warning/error if the user sets the stereo inputs differently.

Thoughts ? cc: @alex-luxonis  @themarpe  @szabi-luxonis

themarpe replied:
I agree as well - CameraBoardSocket represents the port (with convenient naming of current HW positioning) so IMO this option shouldn't be needed.

alex-luxonis commented on include/depthai-shared/common/Extrinsics.hpp:
```diff
@@ -0,0 +1,24 @@
+    // (x, y, z) pose of destCameraSocket w.r.t currentCameraSocket obtained through calibration
+    Point3f translation;
+    // (x, y, z) pose of destCameraSocket w.r.t currentCameraSocket measured through CAD design
+    Point3f measuredTranslation;
```
Another note on the naming here, does it makes sense to replace `measuredTranslation` with `designTranslation` (or `specTranslation`). Just because when actually _measured_, it could be slightly off due to mounting, even though the comment says exactly what it does.

alex-luxonis commented on include/depthai-shared/common/Extrinsics.hpp:
```diff
@@ -0,0 +1,24 @@
+    Point3f translation;
+    // (x, y, z) pose of destCameraSocket w.r.t currentCameraSocket measured through CAD design
+    Point3f measuredTranslation;
+    CameraBoardSocket toCameraSocket = CameraBoardSocket::AUTO;
+    /** FIXME(sachin): Why does fromCameraSocket needs to be vector? fromCameraSockets
+     * is not part of eeprom write. it just enables in linking between cameras after eeprom read.
+     */
+    CameraBoardSocket fromCameraSocket = CameraBoardSocket::AUTO;
```
Just to clarify my understanding, here `fromCameraSocket` need to match with the socket from the `EepromData::cameraData` map?
And is the above comment still actual, or can be removed.

saching13 replied:
I am not using `fromCameraSocket` anymore. Will be removing it before merge.

saching13 commented on include/depthai-shared/common/CameraInfo.hpp:
```diff
@@ -0,0 +1,19 @@
+    std::vector<std::vector<float>> intrinsicMatrix;
+    std::vector<float> distortionCoeff;
+    Extrinsics extrinsics;
+    float measuredFovDeg = 0.0f;  // fov in deg
```
good idea. Will change that.

saching13 commented on include/depthai-shared/common/CameraBoardSocket.hpp:
```diff
@@ -10,4 +10,15 @@ namespace dai {
\ No newline at end of file
+}  // namespace dai
+
+namespace std {
+
+template <>
+struct hash<dai::CameraBoardSocket> {
+    std::size_t operator()(const dai::CameraBoardSocket& s) const {
+        using Type = typename underlying_type<dai::CameraBoardSocket>::type;
+        return std::hash<Type>{}(static_cast<Type>(s));
+    }
+};
+}  // namespace std
```
But agree. 
We can change it to min C++14 on cmake and give it a shot on mac and see.

saching13 replied:
```
set(CMAKE_CXX_STANDARD 14)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
```

I tried setting this in CMakeLists.txt on mac and removed the hash but didn't work.

themarpe replied:
Thanks for the test!

alex-luxonis replied:
It does work for me, with this on Mac (printed by cmake):
`-- The CXX compiler identification is AppleClang 12.0.0.12000032`

Tested with: https://github.com/luxonis/depthai-core/commit/ad000a87d197802605df760a08c76c09857e11c8
(note that in `depthai-core/CMakeLists.txt` the C++ standard version is set differently, this may have overwritten CMAKE_CXX_STANDARD.

Also CI is passing: https://github.com/luxonis/depthai-core/actions/runs/893501003

saching13 replied:
Yeah. you are right. it was overwritten. Thanks for crosschecking.

saching13 commented:
Any more suggestions here? 
I am thinking of finalizing this V6 version and update the `gen2_uvc_calib` branch which will be used for arm calibration for rgb-left setup.

saching13 commented:
If all good with the structure. I will go ahead with the merging tomorrow. 
cc: @alex-luxonis  @themarpe  @szabi-luxonis
